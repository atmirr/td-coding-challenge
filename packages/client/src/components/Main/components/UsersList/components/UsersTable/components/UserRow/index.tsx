import React from 'react';
import { Link } from 'react-router-dom';
import { makeStyles } from '@material-ui/core/styles';
import { TableRow, TableCell } from '@material-ui/core';
import LaunchIcon from '@material-ui/icons/Launch';
import { User as UserType } from 'src/components/Main/components/UsersList';

const useStyles = makeStyles(({ palette }) => ({
  detailIcon: {
    color: palette.text.primary,
  },
}));

function UserRow({ id, name, username }: UserType) {
  const classes = useStyles();
  return (
    <TableRow key={id}>
      <TableCell>{id}</TableCell>
      <TableCell>{name}</TableCell>
      <TableCell>{username}</TableCell>
      <TableCell align="center">
        <Link to={`/user/${id}`}>
          <LaunchIcon className={classes.detailIcon} />
        </Link>
      </TableCell>
    </TableRow>
  );
}

export default UserRow;
