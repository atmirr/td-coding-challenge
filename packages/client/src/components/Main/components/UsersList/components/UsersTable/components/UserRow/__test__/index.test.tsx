import React from 'react';
import '@testing-library/jest-dom';
import '@testing-library/jest-dom/extend-expect';
import { Route, MemoryRouter } from 'react-router-dom';
import { screen, render, fireEvent } from '@testing-library/react';
import UserRow from '../index';

const props = {
  id: 1,
  name: 'Sunny Wucker',
  username: 'madisyn.wuckert51',
};

const Wrapper = ({ children }) => (
  <table>
    <tbody>{children}</tbody>
  </table>
);

describe('<UserRow />', () => {
  it('should renders the component', () => {
    const { container } = render(
      <MemoryRouter initialEntries={['/']}>
        <Wrapper>
          <UserRow {...props} />
        </Wrapper>
      </MemoryRouter>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should display the user's name and username when we pass the props", () => {
    render(
      <MemoryRouter initialEntries={['/']}>
        <Wrapper>
          <UserRow {...props} />
        </Wrapper>
      </MemoryRouter>,
    );
    expect(screen.queryAllByText(props.id).length).toBe(1);
    expect(screen.queryAllByText(props.name).length).toBe(1);
    expect(screen.queryAllByText(props.username).length).toBe(1);
  });

  it("should display the action button to the detail's page of the user", () => {
    const { container } = render(
      <MemoryRouter initialEntries={['/']}>
        <Wrapper>
          <UserRow {...props} />
        </Wrapper>
      </MemoryRouter>,
    );
    const detailButton = container.querySelector('a');
    expect(detailButton).toHaveAttribute('href', `/user/${props.id}`);
  });

  it("clicking on the action button should link to the user's detail page", () => {
    let testLocation;
    const { container } = render(
      <MemoryRouter initialEntries={['/']}>
        <Wrapper>
          <UserRow {...props} />
        </Wrapper>
        <Route
          path="*"
          render={({ location }) => {
            testLocation = location;
            return null;
          }}
        />
      </MemoryRouter>,
    );
    const detailButton = container.querySelector('a');
    fireEvent.click(detailButton);
    expect(testLocation.pathname).toBe(`/user/${props.id}`);
  });
});
