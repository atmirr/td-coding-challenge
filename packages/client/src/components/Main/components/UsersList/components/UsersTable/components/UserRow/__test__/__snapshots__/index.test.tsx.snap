// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserRow /> should renders the component 1`] = `
<table>
  <tbody>
    <tr
      class="MuiTableRow-root"
    >
      <td
        class="MuiTableCell-root"
      >
        1
      </td>
      <td
        class="MuiTableCell-root"
      >
        Sunny Wucker
      </td>
      <td
        class="MuiTableCell-root"
      >
        madisyn.wuckert51
      </td>
      <td
        class="MuiTableCell-root MuiTableCell-alignCenter"
      >
        <a
          href="/user/1"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root makeStyles-detailIcon-1"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 19H5V5h7V3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2v-7h-2v7zM14 3v2h3.59l-9.83 9.83 1.41 1.41L19 6.41V10h2V3h-7z"
            />
          </svg>
        </a>
      </td>
    </tr>
  </tbody>
</table>
`;
